doc/cprover-manual/contracts-loops.md

```diff
@@ -0,0 +1,118 @@
+# Loop Contracts
+
+CBMC offers support for loop contracts, which includes three basic clauses:
+- _invariant_ clause for establishing safety properties
+- _decreases_ clause for establishing termination, and
+- _assigns_ clause for declaring the subset of variables that is modifiable in the loop.
+
+These clauses formally describe an abstraction of a loop for the purpose of a proof.
+CBMC also provides a series of built-in constructs
+to aid writing loop contracts (e.g., _history variables_ and _quantifiers_).
+
+## Overview
+
+Consider an implementation of the [binary search algorithm] below.
+
+```c
+#include <assert.h>
+
+int bin_search(int val, int buf[], int size)
+{
+  if(size < 0) return -1;
+
+  long lb = 0, ub = size - 1;
+  long mid = (lb + ub) / 2;
+
+  while(lb <= ub)
+  {
+     if(buf[mid] == val) break;
+     if(buf[mid] < val)
+       lb = mid + 1;
+     else
+       ub = mid - 1;
+     mid = (lb + ub) / 2;
+  }
+  assert(lb > ub || (0 <= mid && mid < size));
+  return lb > ub ? -1 : mid;
+}
+```
+
+The function stores a lower bound `lb` and an upper bound `ub`
+initialized to the bounds on the buffer `buf`, i.e., to `0` and `size-1` respectively.
+In each iteration, the midpoint `mid` is compared against the target value `val`
+and in case of a mismatch either the lower half or the upper half of the buffer is searched recursively.
+A developer might be interested in verifying two high-level properties on the loop on all possible buffers `buf` and values `val`:
+1. an out-of-bound access should never occur (at `buf[mid]` lookup)
+2. the loop must eventually always terminate
+
+To prove the first (memory-safety) property,
+we may declare _loop invariants_ that must be preserved across all loop iterations.
+In this case, two invariant clauses would together imply that `buf[mid]` lookup is always safe.
```

Comment:
Related to the above comment on the invariant only being at certain points in the loop. This is only here because `buf[mid]` only uses the value of `mid` that is safe at entry point, if we calculated `mid` inside the body and then tried to access `buf[mid]` things could go wrong, e.g.:
```
int badfun(int *buf, int size)
{
  int i = 0;
  int mid = i;
  while(i < size)
  __CPROVER_loop_invariant(0 <=i && i <= size)
  __CPROVER_loop_invariant(mid == i)
  {
    mid = rand();
    if(buf[mid] == 0xCAFE)
      return 1;
    i++;
    mid = i;
  }
}
```
Here even though we have an invariant that `mid` is equal to `i` and `i` is safe, inside the loop we can break this.

Reply:
The bug here is not that the implementation (body) doesn't meet the abstraction (contracts), but that the implementation uses unsafe operations. Using a loop invariant wouldn't mask those. In particular, for this case, the pointer checking flags would expose the unsafe memory access. After instrumentation, CBMC would still check an arbitrary iteration and this particular unsafe access would be caught.